File: Dynamic_Programming/3077. Maximum Strength of K Disjoint Subarrays.cpp
/*
3077. Maximum Strength of K Disjoint Subarrays
*/

// TC:O(n*k) SC:O(n*k*2)
class Solution {
    using LL = long long;
public:
    long long maximumStrength(vector<int>& nums, int k) {
        int n = nums.size();
        nums.insert(nums.begin(),-1);
        vector<vector<vector<LL>>> dp(n+1, vector<vector<LL>>(k+1, vector<LL>(2,LLONG_MIN/2)));
        for(int i = 0; i <= n; i++){
            dp[i][0][0] = 0;
            //dp[i][0][1] = 0;
        }

        dp[1][0][0] = 0;

        for(int i = 1; i <= n; i++){
            for(int j = 1; j <= k; j++){
                if(j % 2 == 0){
                    // do not pick
                    dp[i][j][0] = max(dp[i-1][j][0], dp[i-1][j][1]);
                    dp[i][j][1] = max(dp[i-1][j-1][0], dp[i-1][j-1][1]) - (LL)nums[i]*(k+1-j);
                    dp[i][j][1] = max(dp[i][j][1], dp[i-1][j][1] - (LL)nums[i]*(k+1-j));
                }
                else {
                    dp[i][j][0] = max(dp[i-1][j][0], dp[i-1][j][1]);
                    dp[i][j][1] = max(dp[i-1][j-1][0], dp[i-1][j-1][1]) + (LL)nums[i]*(k+1-j);
                    dp[i][j][1] = max(dp[i][j][1], dp[i-1][j][1] + (LL)nums[i]*(k+1-j));
                }
            }
        }


        return max(dp[n][k][0], dp[n][k][1]);
    }
};

/*
(-1)i+1 * sum[i] * (x - i + 1) 越前面數字可以被乘的越大
只要是選取的第奇數個subarray就會是正數 所以他的sum最好也是正數 i從1開始算
必須選k個
the selected subarrays don't need to cover the entire array.
這題題意不清楚 其實只能按照順序的取 那就可以看得出來是區間型的dp
dp[i][j] : the max strength by finding the first j subarrays from nums[1:i]
-> sum[1]*k - sum[2]*(k-1) + sum[3]*(k-2)-...

x x x x x x x i
兩種情況
1. do not pick nums[i] : dp[i][j] = dp[i-1][j];
suppose k% 2 == 0
2. do pick nums[i] : dp[i][j] = dp[i-1][j-1] - nums[i]*(k+1-j); 自成一國
                     dp[i][j] = dp[i-1][j] - nums[i]*(k+1-j);  這裡是想要跟前面的合併但會發現我們的定義沒有說 前面i-1個肯定會用
多＋一個維度
dp[i][j][0] : the max strength by finding the first j subarrays from nums[1:i] and nums[i] is not part of sum[j]
dp[i][j][1] : the max strength by finding the first j subarrays from nums[1:i]and nums[i] is part of sum[j]
*/

File: Dynamic_Programming/3077. Maximum Strength of K Disjoint Subarrays.py
# TC:O(n*k) SC:O(n)
class Solution:
    def maximumStrength(self, nums: List[int], k: int) -> int:
        '''
        k是odd 代表遇到 k = odd 就是+ even 就是- (k - groups + 1)
        k == 0 就回傳 0
        所以改成走到每一個就判斷一下是要加入之前的體系還是自成一格(j + 1)
        f[k][n] 分成i份以後 前j個數字可以達到的最大總和
        f[i][j] = f[i][j-1](不選), max(f[i-1][j-1] + (s[j] - s[L]) * w_i)
        '''
        n = len(nums)
        s = list(accumulate(nums, initial = 0))
        # 空間優化
        f = [0] * (n + 1)
        for i in range(1, k+1):
            # 最少需要有i個這樣之前才能有i-1組 然後還需要考慮後面要組成的組數 (k - i) 組 要剩下這麼多  n - (k - i) + 1
            pre = f[i-1]
            f[i-1] = mx = -inf
            w = (k - i + 1) * (1 if i % 2 else -1)
            for j in range(i, n - k + i + 1):
                mx = max(mx, pre - s[j-1] * w) # 主要推導公式的時候發現 (s[j] - s[L]) * w_i 可以拆出來
                pre = f[j]
                f[j] = max(f[j-1], s[j] * w + mx)
        return f[n]
        '''
        f = [[0] * (n + 1) for _ in range(k + 1)]
        
        for i in range(1, k+1):
            f[i][i-1] = mx = -inf
            # 最少需要有i個這樣之前才能有i-1組 然後還需要考慮後面要組成的組數 (k - i) 組 要剩下這麼多  n - (k - i) + 1
            w = (k - i + 1) * (1 if i % 2 else -1)
            for j in range(i, n - k + i + 1):
                mx = max(mx, f[i-1][j-1] - s[j-1] * w) # 主要推導公式的時候發現 (s[j] - s[L]) * w_i 可以拆出來
                f[i][j] = max(f[i][j-1], s[j] * w + mx)
        return f[k][n]
        '''
                
        


